### DISCUSSION TITLE
Configuration replacement not working as expected

### DISCUSSION THREAD

**Discussion by mapoallegro** [vladimir-bukhtoyarov/bucket4j#481] [Q&A] [answered]
Hi,

I have a problem with configuration replacement feature.
The replacement is working only when bucket is persisted in Redis, after TTL time (bucket removed from Redis) the configuration is reverted to original one.

Maybe I'm missing something or I made some wrong assumptions.

But from the beginning:

language: **Koltin**
version of bucket4j: **8.10.1**
redis integration using: **LettuceBasedProxyManager**

Below is my code.
I keep my buckets in `ConcurrentHashMap<String, Bucket>()` because I need different configuration for every key/name.

**lettuceBasedProxyManager:** 
```kotlin
@Bean("lettuceProxyManager")
    fun lettuceBasedProxyManager(redisClient: RedisClient): LettuceBasedProxyManager<String> {
        val redisConnection = redisClient.connect(RedisCodec.of(UTF8, INSTANCE))
        return LettuceBasedProxyManager.builderFor(redisConnection)
            .withClientSideConfig(
                ClientSideConfig.getDefault()
                    .withExpirationAfterWriteStrategy(
                        ExpirationAfterWriteStrategy
                            .basedOnTimeForRefillingBucketUpToMax(10.seconds.toJavaDuration()),
                    ),
            )
            .build()
    }
```

**My Component class:**
```Kotlin
@Component
internal class RateLimitExampleOperations(
    private val rateStorage: RateStorage,
    @Qualifier("lettuceProxyManager")
    private val proxyManager: LettuceBasedProxyManager<String>,
) {
    private val rateLimitBuckets = ConcurrentHashMap<String, Bucket>()

    fun consumeToken(name: String): ConsumptionProbe {
        return touchRateLimitBucket(name)
            .tryConsumeAndReturnRemaining(1)
    }

    fun reloadRateLimit(name: String) {
        val rateDto = rateStorage.findByNameOrThrow(name)
        touchRateLimitBucket(name).replaceConfiguration(
            createConfiguration(rateDto).get(),
            PROPORTIONALLY,
        )
    }

    private fun touchRateLimitBucket(name: String): Bucket {
        return rateLimitBuckets.getOrPut(name) {
            val rateDto = rateStorage.findByNameOrThrow(name)
            createRateLimitBucket(rateDto)
        }
    }

    private fun createRateLimitBucket(rateDto: RateDto): Bucket {
        return proxyManager.builder()
            .build(
                rateDto.name,
                createConfiguration(rateDto),
            )
    }

    private fun createConfiguration(rateDto: RateDto): Supplier<BucketConfiguration> {
        return Supplier<BucketConfiguration> {
            createBucketConfiguration(
                tokens = rateDto.limit.toLong(),
                refillTokens = 100,
            )
        }
    }

    private fun createBucketConfiguration(
        tokens: Long,
        refillTokens: Long,
    ): BucketConfiguration {
        return BucketConfiguration.builder()
            .addLimit { limit ->
                limit
                    .capacity(tokens)
                    .refillGreedy(
                        refillTokens,
                        60.seconds.toJavaDuration(),
                    )
                    .initialTokens(tokens)
            }
            .build()
    }
}
```

The `consumeToken` method is responsible for consuming and `reloadRateLimit` for reloading bucket configuration.

I was trying also https://bucket4j.com/8.10.1/toc.html#implicit-configuration-replacement approach but without success.

I thought that when bucket is persisted (saved in Redis) and I execute _ replaceConfiguration_ on it that will update the base configuration of bucket.

**logs:**

```
15:01:04.404 DEBUG p.a.o.d.w.r.Limiter - Token consumed from bucket for name=MAPO, tokens left=10
// UPDATE ON BUCKET set limit = 1
15:01:12.179 DEBUG p.a.o.d.w.r.Limiter - Token consumed from bucket for name=MAPO, tokens left=0
15:01:15.492 DEBUG p.a.o.d.w.r.Limiter - Token consumed from bucket for name=MAPO, tokens left=0
// waiting 10s for TTL to execute
15:01:27.276 DEBUG p.a.o.d.w.r.Limiter - Token consumed from bucket for name=MAPO, tokens left=10
```

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @mapoallegro 

## Explanation

> The replacement is working only when bucket is persisted in Redis, after TTL time (bucket removed from Redis) the configuration is reverted to original one.

The explanation is simple:
1. Config always stored together with bucket state.
2. After replaceConfigureation had called new config was stored in Redis.
3. After expiration happen, both config and bucket state were removed by Redis itself.
4. On next tryConsume Bucket4j had detected the lack of bucket, then it asked config supplier for configuration.
5. Because your configSupplier is constant, it had returned legacy config, so bucket was recreated with legacy config.

## How to fix? 
First of all, you use deprecated method, `build` with constant configuration that [was deprecated](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L123) a halh year ago, to avoid pitfalls similar that your described there. You should switch to [build(K key, Supplier<BucketConfiguration> configurationSupplier)](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L106), and your supplier should be implemented in way that provides most actual configuration.

**Reply by mapoallegro** [reply]
Hi, thanks for quick reply.

I'm using latest version of build. Method `createConfiguration` returning `Supplier` object.

   ```kotlin
 private fun createConfiguration(rateDto: RateDto): Supplier<BucketConfiguration> {
        return Supplier<BucketConfiguration> {
            createBucketConfiguration(
                tokens = rateDto.limit.toLong(),
                refillTokens = 100,
            )
        }
    }
```

So to be clear. I should read actual config (for example from database) inside Supplier code.

Like this:

```kotlin
    private fun createConfiguration(name: String): Supplier<BucketConfiguration> {
        return Supplier<BucketConfiguration> {
            // read actual rateDto for given name from database
            val rateDto = readLimitFromDatabase(name) 
            createBucketConfiguration(
                tokens = rateDto.limit.toLong(),
                refillTokens = 100,
            )
        }
    }
```

And then `reloadRateLimit()` method will reload config for persisted **Bucket** and above change in `createConfiguration(name: String)` will always init bucket with right config in other cases.

**Reply by vladimir-bukhtoyarov** [reply]
> I'm using latest version of build. Method createConfiguration returning Supplier object.

It is not neccesary to update to latest. It nothing wrong to stay at previous version, just switch to proper variant of `build`.

> So to be clear. I should read actual config (for example from database) inside Supplier code.

Yes, you understand correcly. Supplier is called if bucket was never persisted before or was expired, so even reading from slow database should not be a problem, because it should not happen frequently.

**Reply by mapoallegro** [reply]
@vladimir-bukhtoyarov thanks one more time 🙌

Maybe You can add somewhere in documentation how to properly create `BucketConfiguration` using `Supplier` for such cases as mine when we always want to have most actual parameters/configuration for limits/buckets.

Or maybe it's already there and I just missed it ;)

**Reply by vladimir-bukhtoyarov** [reply]
Yes, I am going to update Javadocs and References according to this topic.
